opm/autodiff/BlackoilModel.hpp

```diff
@@ -0,0 +1,435 @@
+                      const Wells*                    wells,
+                      const NewtonIterationBlackoilInterface& linsolver,
+                      const bool has_disgas,
+                      const bool has_vapoil,
+                      const bool terminal_output);
+
+        /// \brief Set threshold pressures that prevent or reduce flow.
+        /// This prevents flow across faces if the potential
+        /// difference is less than the threshold. If the potential
+        /// difference is greater, the threshold value is subtracted
+        /// before calculating flow. This is treated symmetrically, so
+        /// flow is prevented or reduced in both directions equally.
+        /// \param[in]  threshold_pressures_by_face   array of size equal to the number of faces
+        ///                                   of the grid passed in the constructor.
+        void setThresholdPressures(const std::vector<double>& threshold_pressures_by_face);
+
+        /// Called once before each time step.
+        /// \param[in] dt                     time step size
+        /// \param[in, out] reservoir_state   reservoir state variables
+        /// \param[in, out] well_state        well state variables
+        void prepareStep(const double dt,
+                         ReservoirState& reservoir_state,
+                         WellState& well_state);
+
+        /// Called once after each time step.
+        /// In this class, this function does nothing.
+        /// \param[in] dt                     time step size
+        /// \param[in, out] reservoir_state   reservoir state variables
+        /// \param[in, out] well_state        well state variables
+        void afterStep(const double dt,
+                       ReservoirState& reservoir_state,
+                       WellState& well_state);
+
+        /// Assemble the residual and Jacobian of the nonlinear system.
+        /// \param[in]      reservoir_state   reservoir state variables
+        /// \param[in, out] well_state        well state variables
+        /// \param[in]      initial_assembly  pass true if this is the first call to assemble() in this timestep
+        void assemble(const BlackoilState& reservoir_state,
+                      WellStateFullyImplicitBlackoil& well_state,
+                      const bool initial_assembly);
+
+        /// \brief Compute the residual norms of the mass balance for each phase,
+        /// the well flux, and the well equation.
+        /// \return a vector that contains for each phase the norm of the mass balance
+        /// and afterwards the norm of the residual of the well flux and the well equation.
+        std::vector<double> computeResidualNorms() const;
+
+        /// The size (number of unknowns) of the nonlinear system of equations.
+        int sizeNonLinear() const;
+
+        /// Number of linear iterations used in last call to solveJacobianSystem().
+        int linearIterationsLastSolve() const;
```

Comment:
While not used currently, it makes sense to make a function `linearIterationsLastStep()` available in the NewtonSolver class. However I do not think it makes sense to add a *per iteration* report, since you cannot interact with the NewtonSolver in between iterations. Agree?